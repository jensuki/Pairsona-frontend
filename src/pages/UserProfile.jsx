import { useState, useEffect, useContext } from 'react'
import { useParams, useNavigate, Link } from 'react-router-dom';
import UserContext from '../context/UserContext';
import Api from '../helpers/api';
import getProfilePic from '../helpers/getProfilePic';
import Loading from '../components/Loading';
import '../styles/UserProfile.css';

/** component to display user profile cards for either:
 * - the current logged-in user
 * - another user's profile (based on URL parameter)
 *  includes functionality to handle connection requests (send, cancel, accept).
 */

const UserProfile = () => {
    const { username: paramUsername } = useParams(); // get username from url
    const { currUser } = useContext(UserContext);
    const navigate = useNavigate();

    const [user, setUser] = useState(null);
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true); // use local state
    const [connectionStatus, setConnectionStatus] = useState({
        type: null, // 'sent' received' 'connected' or null
        connectionId: null
    });

    // determine which user profile to load
    const username = paramUsername || currUser?.username;

    // fetch user profile + connection status on mount
    useEffect(() => {

        const fetchUserData = async () => {
            try {
                setLoading(true); // start

                const fetchedUser = await Api.getUserProfile(username);
                if (!fetchedUser) {
                    setError('User not found');
                    return;
                }
                setUser(fetchedUser);

                // fetch connection data
                const [sentRequests, pendingRequests, confirmedConnections] = await Promise.all([
                    Api.getSentRequests(),
                    Api.getPendingRequests(),
                    Api.getConnections()
                ]);

                // determine if the current user sent a request to this profile
                const sent = sentRequests.find(req => req.username === fetchedUser.username);

                // determine if the current user received a request from this profile
                const received = pendingRequests.find(req => req.username === fetchedUser.username);

                // determine if the current user is already connected to this profile
                const connected = confirmedConnections.find(conn => conn.username === fetchedUser.username);

                // update connectionStatus based on: connected > received > sent > null
                if (connected) {
                    setConnectionStatus({
                        type: 'connected',
                        connectionId: connected.connectionId,
                    });
                } else if (received) {
                    setConnectionStatus({
                        type: 'received',
                        connectionId: received.connectionId,
                    });
                } else if (sent) {
                    setConnectionStatus({
                        type: 'sent',
                        connectionId: sent.connectionId,
                    });
                } else {
                    setConnectionStatus({
                        type: null,
                        connectionId: null,
                    });
                }
            } catch (err) {
                console.error(err);
                setError('Error fetching user data.');
            } finally {
                setLoading(false);
            }
        };
        fetchUserData();
    }, [username]);

    // handle sending or cancelling a connection request
    const handleConnectionRequest = async () => {
        try {
            setError(null);

            if (connectionStatus.type === 'sent') {
                // if 'sent' -> cancel request and set to null
                await Api.cancelConnectionRequest(connectionStatus.connectionId);
                setConnectionStatus({
                    type: null,
                    connectionId: null
                });
            } else {
                const res = await Api.sendConnectionRequest(username);
                setConnectionStatus({ type: 'sent', connectionId: res.data.id });
            }
        } catch (err) {
            setError(err.message || "Error sending or canceling request");
        }
    };

    // handle accepting a received connection request
    const handleAcceptRequest = async () => {
        try {
            setError(null);
            if (connectionStatus.type === 'received') {
                // accept the received connection request
                await Api.acceptConnectionRequest(connectionStatus.connectionId);
                setConnectionStatus({
                    type: 'connected',
                    connectionId: null
                });
                navigate('/connections'); // redirect to connections page
            }
        } catch (err) {
            setError(err.message || 'Error accepting connection request');
        }
    };

    if (loading) return <Loading />
    if (error) return <p className="error">{error}</p>

    return (
        <div className="UserProfile">

            {/* user profile card image */}
            <div className="card">
                <h1>{user.firstName} {user.lastName}</h1>
                <img
                    src={getProfilePic(user.profilePic)}
                    alt={`${user.firstName} ${user.lastName}`}
                    className="profile-pic"
                    loading="lazy"
                />
                <hr></hr>
                {/* user profile card details */}
                <p>
                    MBTI: <strong>{user.mbti || <Link to="/quiz">Go to quiz!</Link>}</strong>
                </p>
                <p>{new Date(user.birthDate).toLocaleDateString('en-US')}</p>
                <p>{user.location}</p>
                <p>{user.email}</p>
                {user.bio && <p><strong>Bio:</strong> {user.bio}</p>}

                {/* render buttons based on connection status*/}
                {currUser?.username !== user.username && (
                    <>
                        {connectionStatus.type === 'sent' && (
                            <button className="btn btn-cancel" onClick={handleConnectionRequest}>
                                Cancel Connection Request
                            </button>
                        )}
                        {connectionStatus.type === 'received' && (
                            <button className="btn btn-accept" onClick={handleAcceptRequest}>
                                Accept Connection Request
                            </button>
                        )}
                        {connectionStatus.type === 'connected' && (
                            <div className="connected-options">
                                <button className="btn btn-connected" disabled>
                                    Connected
                                </button>
                                <button className="btn btn-chat" onClick={() => navigate(`/messages/${username}`)}>
                                    Chat with {user.firstName}
                                </button>
                            </div>
                        )}
                        {connectionStatus.type === null && (
                            <button className="btn btn-connect" onClick={handleConnectionRequest}>
                                Connect with {username}
                            </button>
                        )}
                    </>
                )}
                {/* display edit profile button if curr user */}
                {currUser?.username === user.username && (
                    <Link to="/profile/edit">
                        <button>Edit Profile</button>
                    </Link>
                )}
            </div>
            {/* mbti information */}
            <div className="mbti-info" data-aos="fade-left">
                {user.mbtiDetails && (
                    <div className="card card-right">
                        <p><strong>{user.mbtiDetails.title}</strong></p>
                        <p>{user.mbtiDetails.description}</p>
                        <a href={user.mbtiDetails.site} target="_blank" rel="noreferrer">
                            Learn more about {user.mbti}
                        </a>
                    </div>
                )}
            </div>
        </div>
    );
}

export default UserProfile;