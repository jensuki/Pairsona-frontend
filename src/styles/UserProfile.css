.UserProfile {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1rem;
    max-width: var(--max-width-md);
    margin: 40px auto;
    padding: 0;
    position: relative;
}

.UserProfile .card {
    margin: 10px auto;
}

.UserProfile h1 {
    margin-top: 0;
}

.UserProfile .btn-cancel {
    background: rgb(218, 122, 105);

    &:hover {
        background: rgb(176, 98, 84);
    }
}

.btn-accept {
    background: rgb(82, 153, 200);

    &:hover {
        background: rgb(67, 120, 155);
    }
}

.btn-connected {
    background: rgb(51, 82, 165);

    &:hover {
        background: rgb(40, 69, 150);
        cursor: not-allowed;
    }
}

/* user card connection status */

.connected-options {
    display: flex;
    gap: 10px;
}

/* matched user mbti detail */

.card-right {
    border: 2px solid rgb(246, 196, 204);
}

/* span side by side on large screens */

@media (min-width: 768px) {
    .UserProfile {
        flex-direction: row;
        align-items: flex-start;
        max-width: 700px;
        justify-content: center;
    }

    .card-right {
        width: auto;
        flex: 1;
    }
}