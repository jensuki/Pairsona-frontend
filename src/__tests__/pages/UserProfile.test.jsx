import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import UserProfile from '../../pages/UserProfile';
import Api from '../../helpers/api';
import { describe, test, expect, vi, beforeEach } from 'vitest';
import { demoUser, demoConnections, demoConnectionRequests } from '../../test_common/mockData';

// mock api funcs
vi.mock('../../helpers/api', () => ({
    default: {
        getUserProfile: vi.fn(),
        getSentRequests: vi.fn(),
        getPendingRequests: vi.fn(),
        getConnections: vi.fn(),
        sendConnectionRequest: vi.fn(),
        cancelConnectionRequest: vi.fn(),
        acceptConnectionRequest: vi.fn()
    }
}))

const renderUserProfile = () => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={demoUser}>
                <UserProfile />
            </UserProvider>
        </MemoryRouter>
    )
}

describe('UserProfile component', () => {
    beforeEach(() => {
        vi.clearAllMocks();

        // mock responses
        Api.getUserProfile.mockResolvedValue(demoUser); // mock demouser profile data
        Api.getSentRequests.mockResolvedValue([]);
        Api.getPendingRequests.mockResolvedValue(demoConnectionRequests);
        Api.getConnections.mockResolvedValue(demoConnections);
    })

    test('renders without crashing and displays curr user info', async () => {
        renderUserProfile();

        // wait for users name to appear in the dom
        await waitFor(() => {
            expect(screen.getByText(`${demoUser.firstName} ${demoUser.lastName}`)).toBeInTheDocument();

        })
        // expect other fields to be present
        const formattedBirthDate = new Date(demoUser.birthDate).toLocaleDateString('en-US');
        expect(screen.getByText(formattedBirthDate)).toBeInTheDocument()
        // expect edit profile since viewing logged in users own profile
        expect(screen.getByText('Edit Profile')).toBeInTheDocument();
    })

    test('matches snapshot', async () => {
        const { asFragment } = renderUserProfile();

        // Wait for the user's name to appear (indicating the component has finished loading)
        await waitFor(() => {
            expect(screen.getByText(`${demoUser.firstName} ${demoUser.lastName}`)).toBeInTheDocument();
        });

        // Capture the snapshot after the component has finished rendering
        expect(asFragment()).toMatchSnapshot();
    });

    test('displays loading spinner while fetching data', async () => {
        const { getByRole } = renderUserProfile();

        await waitFor(() => {
            expect(getByRole('status')).toBeInTheDocument();
        })
    });

    test('renders cancel connection button if connection is sent', async () => {
        // use first user from mock connections
        const otherUser = demoConnections[0];

        // mock api response for other users profile and connection data
        Api.getUserProfile.mockResolvedValueOnce(otherUser); // mock other profile being viewed
        Api.getSentRequests.mockResolvedValueOnce([
            {
                username: otherUser.username,
                connectionId: 1
            }
        ]);

        Api.getConnections.mockResolvedValueOnce([]); // no confirmed connection with this profile

        renderUserProfile(); //

        // wait for 'cancel connection request' button to appear on other users profile
        await waitFor(() => {
            expect(screen.getByText('Cancel Connection Request')).toBeInTheDocument();
        })
    })

    test('renders accept connection button for a received connection request', async () => {
        // use first user from mock connections
        const otherUser = demoConnections[0];

        // mock api response
        Api.getUserProfile.mockResolvedValueOnce(otherUser); // mock other user's profile
        Api.getPendingRequests.mockResolvedValueOnce([
            {
                connectionId: 1,
                username: otherUser.username,
                sender: otherUser.username, // sender is otherUser
                recipient: demoUser.username, // recipient is the curr user
                status: 'pending',
            },
        ]);
        Api.getConnections.mockResolvedValueOnce([]); // no confirmed connections between the two

        renderUserProfile();

        // wait for 'accept connection request' button to appear
        await waitFor(() => {
            expect(screen.getByText('Accept Connection Request')).toBeInTheDocument();
        });

        // click accept button
        const acceptButton = screen.getByText('Accept Connection Request');
        fireEvent.click(acceptButton);

        // ensure correct connection id was accepted
        await waitFor(() => {
            expect(Api.acceptConnectionRequest).toHaveBeenCalledWith(1);
        });
    });

    test('renders connected options if users are connected', async () => {
        const otherUser = demoConnections[0];

        Api.getUserProfile.mockResolvedValueOnce(otherUser); // mock other users profile
        Api.getConnections.mockResolvedValueOnce([
            {
                connectionId: 1,
                username: otherUser.username, // mock a confirmed connection with the other user
            },
        ]);

        renderUserProfile();

        // wait for 'connected' button to appear
        await waitFor(() => {
            expect(screen.getByText('Connected')).toBeInTheDocument();
            expect(screen.getByText(`Chat with ${otherUser.firstName}`)).toBeInTheDocument();
        });
    });
})
