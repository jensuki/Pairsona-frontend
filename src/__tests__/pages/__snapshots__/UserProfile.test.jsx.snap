// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserProfile component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="UserProfile"
  >
    <div
      class="card"
    >
      <h1>
        Test User
      </h1>
      <img
        alt="Test User"
        class="profile-pic"
        loading="lazy"
        src="/assets/default_pic.png"
      />
      <hr />
      <p>
        MBTI: 
        <strong>
          INFJ
        </strong>
      </p>
      <p>
        12/31/1999
      </p>
      <p>
        New York, NY
      </p>
      <p>
        testuser@example.com
      </p>
      <p>
        <strong>
          Bio:
        </strong>
         Loves personality tests
      </p>
      <a
        data-discover="true"
        href="/profile/edit"
      >
        <button>
          Edit Profile
        </button>
      </a>
    </div>
    <div
      class="mbti-info"
      data-aos="fade-left"
    />
  </div>
</DocumentFragment>
`;
